text/0477-policy-validation.md

```diff
@@ -0,0 +1,553 @@
+      },
+    }),
+  ],
+});
+
+// only apply to a particular stage
+const prodStage = new Stage(app, 'ProdStage', {
+  validationPlugins: [...],
+});
+```
+
+#### Alternative
+
+```ts
+const app = new App();
+Validations.of(app).add(new CfnGuardValidator());
+```
+
+When you synthesize the CDK app the validator plugins will be called and the results will be printed.
+
+```text
+$ cdk synth
+
+Validation Report (CfnGuardValidator)
+-------------------------------------
+
+(Summary)
+
+╔═══════════╤════════════════════════╗
+║ Status    │ failure                ║
+╟───────────┼────────────────────────╢
+║ Plugin    │ CfnGuardValidator      ║
+╟───────────┼────────────────────────╢
+║ Version   │ 1.3.4                  ║
+╟───────────┼────────────────────────╢
+║ Customize │ (policy location.      ║
+║ policy    │                        ║
+╚═══════════╧════════════════════════╝
+
+(Violations)
+
+Ensure S3 Buckets are encrypted with a KMS CMK (2 occurrences)
+
+  Occurrences:
+  
+    - Construct Path: cdk-app/MyStack/Bucket/Resource
+    - Stack Template Path: ./cdk.out/MyStack.template.json
+    - Creation Stack: [
+        'new Bucket (/home/hallcor/tmp/cdk-tmp-app/node_modules/aws-cdk-lib/aws-s3/lib/bucket.js:1:12889)',
+        'new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)',
+      ]
+    - Resource Name: my-bucket
+    - Locations:
+      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm
+
+  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
+  How to fix:
+    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
+    > Add to construct properties for `cdk-app/MyStack/Bucket`
+      `encryption: BucketEncryption.KMS`
+
+  
+Validation failed. See above reports for details
+```
+
+### Adding Exemptions/Suppressions
+
+> Prior art: [cdk-nag](https://github.com/cdklabs/cdk-nag).
+
+There may be cases where you would like to suppress a certain rule. To do so you must specify the rule to suppress and
+the reason for the suppression.
+
+A suppression can be added for all resources under a construct scope. For
+example to add suppressions for an entire stack.
+
+```ts
+import { Stack, Validations } from 'aws-cdk-lib';
+
+const stack = new Stack(app, 'DevStack');
+Validations.of(stack).addSuppression({
+  rule: 'S3BucketEncryption'
+  reason: 'Dev environment buckets do not have to be encrypted'
+});
+```
+
+They can also be added for specific resources.
+
+```ts
+class MyStack extends Stack {
+  constructor(scope: Construct, id: string, props?: StackProps) {
+    const bucket = new Bucket(this, 'MyBucket');
+    Validations.of(bucket).addSuppression({
+      rule: 'S3BucketEncryption',
+      reason: 'This bucket does not require encryption because xyz',
+    });
+  }
+}
+```
+
+Each plugin will determine how the suppressions are handled for the specific tool used, but the CDK will also synthesize
```

Comment:
I think the plugins have to access suppressions via the template. The rules that are run locally should be the same rules that are eventually run as part of CFN hooks so it needs to be the rules themselves that have the logic for identifying suppressions and not the plugin.

I think the suppressions need to be written to each individual resource. Maybe instead of the metadata we should use tags instead since those will be available in the template as well as the resource itself (and available to Config). 

Since the plugins run after the template is synthesized I don't think we can have the plugin report back anything since it is too late to modify the template at this point.

> And again going back to my previous question, how will plugins know that a specific suppression relates to them and not to some other plugin?

I don't think this matters. If there are multiple plugins that both have the same rule I think a single suppression would be fine.